chord_metadata_service/restapi/middleware.py

```diff
@@ -1,86 +1,104 @@
-            # the signature, gotta find a proper way to transmit vault's public key
-            decoded_jwt = jwt.decode(authz_jwt, verify=False)
-
-            if 'permissions' in decoded_jwt:
-                permissions = decoded_jwt.get('permissions')
-
-                # TODO: I am assuming the permission object is a simple dict such as
-                # {"DATASET_NAME": 4}
-                # TODO: would make sense to package this feature, the reading of the
-                # permission object in a lib, since we'll repeat that stuff over many services
-                if dataset in permissions:
-                    return permissions[dataset]
-                else:
-                    return
-
-    def query_opa(self, request, access_level):
-        if not settings.CANDIG_OPA_URL:
-            return False
-
-        params = {
-            "input": {
-                "path": request.path,
-                "method": request.method,
-                "access_level": int(access_level)
+        """
+        You may incorporate any data source for dataset authorization.
+
+        Note, if no datasets are authorized, you MUST set it to 'NO_DATASETS_AUTHORIZED'.
+
+        If authorized datasets exist, set it to a comma-separated string with titles of authorized datasets.
+
+        For example, if datasets d100 and d200 are authorized, you should set
+        self.authorize_datasets = 'd100,d200'
+        """
+        
+        if settings.CANDIG_AUTHORIZATION == 'OPA' and request.method == 'GET'\
+            and any(re.match(path_re, request.path) for path_re in self.authorized_paths):
+            if settings.CACHE_TIME != 0:
+                error_response = {
+                    "error": "cache time needs to be zero to be secure"
+                }
+                response = HttpResponseServerError(json.dumps(error_response))
+                response["Content-Type"] = "application/json"
+                return response
+            token = self.get_opa_token_from_request(request.headers)
+            opa_res_datasets = self.get_opa_res(token, request.path, request.method)
+            if len(opa_res_datasets) == 0:
+                self.authorize_datasets = 'NO_DATASETS_AUTHORIZED'
+            elif type(opa_res_datasets) == tuple and opa_res_datasets[0] == "error": # error response
+                return opa_res_datasets[1]
+            else:
+                self.authorize_datasets = ",".join(opa_res_datasets)
+            request.GET = request.GET.copy() # Make request.GET mutable
+            request.GET.update({'authorized_datasets': self.authorize_datasets})
+            response = self.get_response(request)
+            return response
+        response = self.get_response(request)
+        return response
+
+    def get_opa_token_from_request(self, headers):
+        """
+        Extracts token from request's header X-CANDIG-LOCAL-OIDC
+        """
+        token = headers.get("X-CANDIG-LOCAL-OIDC")
```

Comment:
I am not sure if the authorization token to Katsu is included as a header `X-CANDIG-LOCAL_OIDC`, my understanding is that if Katsu is registered as a Tyk API, its token is actually also included as a `Bearer {token}` of the `Authorization` Header, need to check with Jonathan